node/rate_limiter.js

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var assert = require('assert');
+
+var DEFAULT_SERVICE_RPS_LIMIT = 100;
+var DEFAULT_TOTAL_RPS_LIMIT = 1000;
+var DEFAULT_BUCKET_NUMBER = 20;
+
+function RateLimiterCounter(options) {
+    if (!(this instanceof RateLimiterCounter)) {
+        return new RateLimiterCounter(options);
+    }
+
+    var self = this;
+    self.index = 0;
+    self.rps = 0;
+    self.numOfBuckets = options.numOfBuckets;
+    self.buckets = [];
+    self.buckets.length = self.numOfBuckets;
+
+    // self.buckets is read/written in refresh,
+    // where read is always after write on a bucket.
+    self.buckets[0] = 0;
+    self.rpsLimit = options.rpsLimit;
+}
+
+RateLimiterCounter.prototype.refresh =
+function refresh() {
+    var self = this;
+
+    // update the sliding window
+    var next = (self.index + 1) % self.numOfBuckets;
+    if (self.buckets[next]) {
+        // offset the bucket being moved out
+        self.rps -= self.buckets[next];
+    }
+
+    assert(self.rps >= 0, 'rps should always be larger equal to 0');
+    self.index = next;
+    self.buckets[self.index] = 0;
+};
+
+RateLimiterCounter.prototype.increment =
+function increment() {
+    var self = this;
+    self.buckets[self.index] += 1;
+    self.rps += 1;
+};
+
+function RateLimiter(options) {
```

Comment:
I am going to say keep this design for now, but note that it limits the flexibility of the interface by coupling the logic of rate limiting to an index of rate limiters by service and by total. Alternately, you could have a RateLimiter instance per service (indexed externally by service name) and another rate limiter instance for the total. This would make it possible to use the rate limiter logic for other scenarios, e.g., indexed by host:port, which we may eventually need to penalize ephemeral clients, or just instantiated independently for testing, benchmarking, and A/B comparisons in isolation.

As @Raynos has observed in other places, even with an instance of a RateLimiter for each service, we would still want a collection of RateLimiters indexed by service to consolidate all of the book keeping under a single timer (as you have with this design).

I say to not make any changes yet because this design may prove better because it has fly-weight rate limiters for each service, reducing memory usage and possibly increasing cache locality. A more flexible design may have a performance cost.

That said, I insist that there is still room to experiment with an alternate design that eliminates the timers and reduces memory usage by using a single statistic per rate limiter and adjusting that statistic on the leading or trailing edge of some sample of requests.